File: 64-minimum-path-sum/64-minimum-path-sum.cpp
class Solution {
private:
    int f(int m, int n, vector<vector<int>> &grid, vector<vector<int>> &dp) {
        if(m == 0 and n == 0) return grid[0][0];
        if(m < 0 or n < 0) return 1e7;
        if(dp[m][n] != -1) return dp[m][n];
        int left = grid[m][n] + f(m, n - 1, grid, dp);
        int up = grid[m][n] + f(m - 1, n, grid, dp);
        return dp[m][n] = min(left, up);
    }
public:
    int minPathSum(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        vector<vector<int>> dp(m, vector<int> (n, -1));
        return f(m - 1, n - 1, grid, dp);
    }
};

File: 64-minimum-path-sum/64-minimum-path-sum.java
class Solution {
    public int minPathSum(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        int[] dp = new int[n];
        for(int i = 0; i < m; ++i) {
            int[] temp = new int[n];
            for(int j = 0; j < n; ++j) {
                if(i == 0 && j == 0) temp[j] = grid[0][0];
                else {
                    int left = 0, up = 0;
                    
                    if(j > 0) left += grid[i][j] + temp[j - 1];
                    else left += 1e7;
                    
                    if(i > 0) up += grid[i][j] + dp[j];
                    else up += 1e7;
                    
                    temp[j] = Math.min(left, up);
                }
            }
            dp = temp;
        }
        return dp[n - 1];
    }
}